src/gluonts/dataset/pandas.py

```diff
@@ -84,71 +84,63 @@ class PandasDataset:
     timestamp: Optional[str] = None
     freq: Optional[str] = None
     static_features: pd.DataFrame = pd.DataFrame()
-    ignore_last_n_targets: int = 0
+    future_length: int = 0
     unchecked: bool = False
     assume_sorted: bool = False
     dtype: Type = np.float32
-    _pairs: Iterable[Tuple[Any, Union[pd.Series, pd.DataFrame]]] = field(
-        init=False
-    )
     _static_reals: pd.DataFrame = field(init=False)
     _static_cats: pd.DataFrame = field(init=False)
 
     def __post_init__(self):
         if isinstance(self.dataframes, dict):
-            self._pairs = self.dataframes.items()
+            pairs = self.dataframes.items()
         elif isinstance(self.dataframes, (pd.Series, pd.DataFrame)):
-            self._pairs = [pair_with_item_id(self.dataframes)]
+            pairs = [(None, self.dataframes)]
         else:
             assert isinstance(self.dataframes, SizedIterable)
-            self._pairs = Map(pair_with_item_id, self.dataframes)
+            pairs = Map(pair_with_item_id, self.dataframes)
 
-        assert isinstance(self._pairs, SizedIterable)
+        self._data_entries = StarMap(self._pair_to_dataentry, pairs)
+
+        if self.feat_dynamic_real is None:
+            self.feat_dynamic_real = pd.DataFrame()
+
+        if self.past_feat_dynamic_real is None:
+            self.past_feat_dynamic_real = pd.DataFrame()
 
         if self.freq is None:
             assert (
                 self.timestamp is None
             ), "You need to provide `freq` along with `timestamp`"
-            self.freq = infer_freq(first(self._pairs)[1].index)
+            self.freq = infer_freq(first(pairs)[1].index)
 
-        self._static_reals = self.static_features.select_dtypes(
-            "number"
-        ).astype(self.dtype)
-        self._static_cats = self.static_features.select_dtypes(
-            "category"
-        ).apply(lambda col: col.cat.codes)
+        self._static_reals = (
+            self.static_features.select_dtypes("number").astype(self.dtype).T
+        )
 
-        self._data_entries = Map(self._pair_to_dataentry, self._pairs)
+        self._static_cats = (
+            self.static_features.select_dtypes("category")
+            .apply(lambda col: col.cat.codes)
+            .T
+        )
 
     @property
     def num_feat_static_cat(self):
-        return len(self._static_cats.columns)
+        return len(self._static_cats)
 
     @property
     def num_feat_static_real(self):
-        return len(self._static_reals.columns)
+        return len(self._static_reals)
 
     @property
     def num_feat_dynamic_real(self):
-        if self.feat_dynamic_real is None:
-            return 0
         return len(self.feat_dynamic_real)
 
     @property
     def num_past_feat_dynamic_real(self):
-        if self.past_feat_dynamic_real is None:
-            return 0
         return len(self.past_feat_dynamic_real)
```

Comment:
These changes work as well if `self.feat_dynamic_real` is forced to be `[]` when found `None` (same for the other)